### PR TITLE
Skip hidden *nix files when populating gamelists.

### PR DESCRIPTION
**Pull request by dankcushions**
Re-do of  #182 - turns out adding options was pretty easy to do blind :)

- Has option in 'OTHER SETTINGS' - 'SHOW HIDDEN FILES' - defaults to false
- Re-wrote function myself.
- Removed folder check - existing ES didn't seem to show hidden directories anyway, so it would never go into my function.
- Requires ES restart to take effect.

Will do one final test tonight.

### CODE REVIEW

File: es-app/src/SystemData.cpp [hunk 1]
```diff
@@ -63,6 +63,23 @@ void SystemData::setIsGameSystemStatus()
 	mIsGameSystem = (mName != "retropie");
 }
 
+// test to see if a file is hidden in *nix (dot-prefixed)
+// could be expanded to check for Windows hidden attribute
+bool isHidden(const fs::path &filePath)
+{
+	if(Settings::getInstance()->getBool("ShowHiddenFiles"))
+		return false;
```
Comment (joolswills) [inline es-app/src/SystemData.cpp]: It doesn't seem logical to me that the check would be in the function. The function should always do the operation - the check should be done on the code that calls the

File: es-app/src/SystemData.cpp [hunk 2]
```diff
@@ -63,6 +63,23 @@ void SystemData::setIsGameSystemStatus()
...
+
+	fs::path::string_type fileName = filePath.filename().string();
+	if(fileName[0] == '.')
+	{
+		LOG(LogWarning) << "File \"" << fileName << "\" is hidden. Ignoring it.";
```
Comment (joolswills) [inline es-app/src/SystemData.cpp]: I wouldn't have thought there needs to be any logging here.

**Comment by hex007** [comment]
@dankcushions Is the option necessary? I can see this as always enabled.

**Comment by joolswills** [comment]
I want it as an option so the old behaviour can be switched to if desired.

**Comment by joolswills** [comment]
Just wondering if there could be a performance hit here with it calling the function and getting the setting value each time like this.

Maybe something like this would suffice (untested).

```
--- a/es-app/src/SystemData.cpp
+++ b/es-app/src/SystemData.cpp
@@ -88,6 +88,7 @@ void SystemData::populateFolder(FileData* folder)
        fs::path filePath;
        std::string extension;
        bool isGame;
+       bool showHidden = Settings::getInstance()->getBool("ShowHiddenFiles");
        for(fs::directory_iterator end, dir(folderPath); dir != end; ++dir)
        {
                filePath = (*dir).path();
@@ -103,7 +104,7 @@ void SystemData::populateFolder(FileData* folder)
                //see issue #75: https://github.com/Aloshi/EmulationStation/issues/75
 
                isGame = false;
-               if(std::find(mEnvData->mSearchExtensions.begin(), mEnvData->mSearchExtensions.end(), extension) != mEnvData->mSearchExtensions.end())
+               if(std::find(mEnvData->mSearchExtensions.begin(), mEnvData->mSearchExtensions.end(), extension) != mEnvData->mSearchExtensions.end() && (showHidden || filePath.filename().string()[0] != '.') )
                {
                        FileData* newGame = new FileData(GAME, filePath.generic_string(), mEnvData, this);
                        folder->addChild(newGame);
```

**Comment by dankcushions** [comment]
I had the same thought :) I think you could even get the value in an earlier function (otherwise I think it unnecessarily re-retrieves it for every folder). at some point it becomes less easy to follow for probably immeasurable performance gain.

i'm out of the country this weekend but will have a look next week.

**Comment by dankcushions** [comment]
i just tried it - from when I see the ES splash screen until it goes away:

- my previous version 2:21
- latest binary 2:22
- jools' suggestion 2:20

looks like any differences are negligible or compiled away. i've squashed everything.

although... one kind of bizarre thing is that if you 'restart emulationstation' it takes a matter of seconds to load back up again, but still respects the 'hidden files' setting if it's changed, so it must be going through everything again? not sure which part takes all the time during a full restart.. weird.

**Comment by pjft** [comment]
While I didn't necessarily test the restart times, I don't expect this change to have had any impact in that :)

I just tested it and it does what it says on the tin - thanks for putting it together!

One small comment, for the sake of usability. If it requires one to restart in order to benefit from it, maybe we could let the user know about it. 

There are several ways we could go about it, and truth be told, I'll leave it to your discretion as to how to make that happen. It could range from a popup saying that it requires a restart, and asking the user if he wants to restart now (this is slightly more complex), to a simple change in the option caption to say "(Requires ES Restart)". 

I'm happy with the last one, truth be told. Just to pre-empt people saying that they changed the option but files are still hidden/visible. :)

Thanks!

**Comment by dankcushions** [comment]
> One small comment, for the sake of usability. If it requires one to restart in order to benefit from it, maybe we could let the user know about it.

i didn't think it was possible to over-engineer this more, but here we are ;)

> a simple change in the option caption to say "(Requires ES Restart)".

i don't agree with adding more stuff to the name as it looks ugly to me and sidesteps the lack of contextual help in ES. bear in mind this option will be set correctly (off) by default and i expect there to be 0 users who will change it anyway. plus they're used to restarting ES to show newly-added files so to me it makes 'sense' that this would require a restart.

i think ES needs contextual help but that's a general issue. 'VRAM limit' & 'Parse Gamelists Only' are proven 'dangerous' options that are (specifically the latter) responsible for many user issues but have no onscreen help. 

also i think we need to be very conservative with adding/expanding options until we allow theming of them as they're already starting to approach the edges (or go over?) low resolution displays (CRT).

outside of contextual help, the only improvement i could accept would be that changing it would trigger a restart/populate folder loop, but that would require more C++ than I have.

sorry for the rant! i agree in principle, but i hope it's not needed :)

**Comment by pjft** [comment]
I thought font size was calculated a proportion of the screen, but I can
see how different aspect ratios can affect that.

Not going to argue with it much - it isn't a blocker in any way, it was
more of a comment. The fact that other options/parts of ES may be missing
things we feel would be useful (like a better explanation) shouldn't
prevent us from doing it better/right in a new instance, if we feel that's
the way to go - otherwise things will never improve. :)

If adding a caption isn't an option, I'm fine with it as is.

Thanks!




- Paulo Tavares

**Comment by joolswills** [comment]
I saw you committed my suggestion, but looks like it was removed when you squashed - just to check was that intentional ?

**Comment by dankcushions** [comment]
yes - I rolled back after I saw that it made no difference to the speed. I feel the current way is more maintainable (for windows support later), if I had to pick one! Don't really mind, though.

**Comment by joolswills** [comment]
it may not have been measurable but I still prefer not getting the config value in the loop each time for example. Also it may make a different on slower devices such as the pi1.

Not sure there would be a problem with hidden files on windows would there ?

**Comment by dankcushions** [comment]
Done :)

> Not sure there would be a problem with hidden files on windows would there ?

no, just that the option wouldn't do anything in windows (or at least, do an unexpected thing by treating . prefixed files as hidden)

**Comment by pjft** [comment]
Oh! Well spotted.

We could add the #ifdef _RPI_ / #endif tags around those, then, as well as
the menu option.

See what was done for OMX player.

**Comment by dankcushions** [comment]
This option wouldn't just apply to RPI, though, but any *nix system (OSX, x86 Ubunto, Raspbian, etc). I saw WIN32 around some other windows things so I have implemented that. Seems to work as normal on my RPI.

**Comment by pjft** [comment]
Of course, you shouldn't trust emails from me at 5:20AM.

That is the correct approach.


- Paulo Tavares

**Comment by joolswills** [comment]
Thanks. Please squash it and I will merge.

**Comment by dankcushions** [comment]
Done :+1:

**Pull request merged**